### DISCUSSION TITLE
How to hide query field value depending on role?

### DISCUSSION THREAD

**Discussion by matart15** [hasura/graphql-engine#9403] [Question]
Lets say 

I have Post table
```ts
model Post {
  id: string;
  title: string;
  detail: string;
  someData: string;
}
```

I have multiple dynamic roles ( with permissions ). 
- Admin : can see `someData` field
- User : can not see `someData` field

My requirement is 
- My UI needs to display all Post fields.
- If role does not have permission to see someData My UI will just display empty.

Ideally i would send this query and depending on permission someData will return null
```gql
query post {
  Post {
    id
    title
    detail
    someData
  }
}
```

If i set Permission on User, It hides `someData` field on query. Which results throwing `someData1 does not exist on post query` error

How do set permission and not Hide `someData` field?

**Comment by matart15** [comment]
If my Roles and permissions are static, I can create 2 separate query and use according to my roles

```gql
query AdminPost {
  Post {
    id
    title
    detail
    someData
  }
}
```

```gql
query UserPost {
  Post {
    id
    title
    detail
  }
}
```

But, My  permissions are set dynamically for users

**Reply by meetzaveri** [reply]
Hey @matart15 

When you open API explorer and see the query, it hides that field because for that specific role, the introspection query execution respects the authz (permissions) layer. 

Introspection query is responsible for showing you the schema structure, generating query and mutations you can use for specific user role. Based on the headers which involves user role, authentication or any custom headers, it checks permissions for each table associated with that role, and then it generates final schema which may hide fields/queries/mutations if role isn't permitted for accessing it.

I hope this makes sense. Thanks for asking!

**Reply by matart15** [reply]
@meetzaveri  Thank you. 

Yes I understand thtat introspection changes depending on permission. 

My problem is My graphql request is static. But my permission is dynamic. So, When one of field's permission dynamically removed my request not pass introspection check

Lets say I setup hasura permission for field
- If `user.permission`'s value includes `Post.someData` string user can show `Post.somedata`

1. `userA.permission` = [`Post.somedata`, `Post.id`, `Post.title`]
2. my UI can show post with this graphql
```gql
query Post {
  Post {
    id
    title
    someData
  }
}
```
3. Then I delete `Post.somedata` permission from `user.permission` array
4. Now same UI gives me error `someData` does not exists on Post. I want null value on someData field.

**Reply by meetzaveri** [reply]
That's the limitation here. GraphQL is strongly typed in nature and introspection returns the schema which does not involve type definition for that field (i.e. `someData` ) in this case. Hence, the error results because of graphql validation.

**Reply by matart15** [reply]
I see. If I want to achive field level permission, hasura is not the tool for me

**Reply by matart15** [reply]
Does that mean hasura have no option to mask fields?

like admin can see user's name and email, but users can see only names.

I think without this function it would be very hard to do any useful application

**Reply by matart15** [reply]
Sorry it sounded little bit aggressive. No bad meaning, just trying to explain my point. Hasura is good tool.

**Reply by meetzaveri** [reply]
No worries, can I know which Hasura version are you using ? 

Also to clarify, this error is not Hasura specific error but rather than related GraphQL spec in general. For reference - https://spec.graphql.org/October2021/#sec-Handling-Field-Errors

Given the fact you decided not to include that column in permission for `user` role, then Hasura will run introspection query which will return with the field being omitted in generated schema and thus when you query, it will result into error (the validation error which graphql throws upon execution)

**Reply by matart15** [reply]
I am using hasura `v2.14.1`

> Also to clarify, this error is not Hasura specific error but rather than related GraphQL spec in general.

Yes I understand. I was just being lazy and looking for out of box solution. I can always create new query or mutation wit remote schema or Hasura action. and return null depending on permission

But creating and managing whole express server just for few fields sounded too much

If i go down this route ( create express graphql server ) there is no merit of using hasura

**Reply by matart15** [reply]
@meetzaveri 

Thank you. I was just wondering If I am missing out some implemented feature. Thanks to you now I know that it can't be done without custom server.

**Reply by leoalves** [reply]
Instead of setting the permission to the field you could:

If `somedata` is a single field you can create a computedField and only return the row value if the role is admin otherwise return null.

If `somedata` is a group of fields I would create a view, hide the view from the root_query, add a relationship between the view and the table and set the permission to the view allowing role admin to return the data and other roles with a permission that would always be false. Something like this:
https://hasura.io/docs/latest/auth/authorization/role-multiple-rules/

**Reply by matart15** [reply]
@leoalves  

For now I created remote schema and created custom query with apolloServer.

I will try https://hasura.io/docs/latest/auth/authorization/role-multiple-rules/ later.

Thank you
